File: python-migration/src/parser.py
"""Markdown parser with front matter extraction."""

import os
import re
from pathlib import Path
from typing import List, Dict, Any
import frontmatter
import markdown
from datetime import datetime

from .types import MarkdownFile, FrontMatter


class MarkdownParser:
    """Parses markdown files and extracts front matter."""
    
    def __init__(self):
        self.md = markdown.Markdown(
            extensions=[
                'codehilite',
                'fenced_code',
                'tables',
                'toc'
            ]
        )
    
    def find_markdown_files(self, input_path: str, pattern: str = "**/*.md") -> List[str]:
        """Find all markdown files in the given directory."""
        path = Path(input_path)
        if not path.exists():
            raise FileNotFoundError(f"Input path does not exist: {input_path}")
        
        # Use glob to find markdown files
        files = list(path.glob(pattern))
        
        # Convert to absolute path strings
        return [str(f.resolve()) for f in files if f.is_file()]
    
    def parse_file(self, file_path: str) -> MarkdownFile:
        """Parse a single markdown file."""
        if not os.path.exists(file_path):
            raise FileNotFoundError(f"File does not exist: {file_path}")
        
        # Load and parse the file with python-frontmatter
        with open(file_path, 'r', encoding='utf-8') as f:
            post = frontmatter.load(f)
        
        # Extract and normalize front matter
        front_matter = self._normalize_front_matter(post.metadata, file_path)
        
        # Convert markdown to HTML
        html = self._convert_to_html(post.content, front_matter)
        
        return MarkdownFile(
            path=file_path,
            front_matter=front_matter,
            content=post.content,
            html=html
        )
    
    def _normalize_front_matter(self, metadata: Dict[str, Any], file_path: str) -> FrontMatter:
        """Normalize front matter data."""
        # Generate slug from filename if not provided
        if not metadata.get('slug'):
            filename = Path(file_path).stem
            metadata['slug'] = self._generate_slug(filename)
        
        # Detect type from path if not specified
        if not metadata.get('type'):
            metadata['type'] = self._detect_type_from_path(file_path)
        
        # Ensure arrays for array fields
        array_fields = ['authors', 'tags', 'categories', 'initiatives', 'guests']
        for field in array_fields:
            if field in metadata and not isinstance(metadata[field], list):
                metadata[field] = [metadata[field]] if metadata[field] else []
        
        # Normalize dates
        date_fields = ['date_published', 'date_updated', 'start_date', 'end_date']
        for field in date_fields:
            if field in metadata:
                metadata[field] = self._normalize_date(metadata[field])
        
        # Set external_id if not provided
        if not metadata.get('external_id'):
            metadata['external_id'] = metadata['slug']
        
        # Create FrontMatter object with only valid fields
        valid_fields = {
            field.name for field in FrontMatter.__dataclass_fields__.values()
        }
        
        filtered_metadata = {
            k: v for k, v in metadata.items() 
            if k in valid_fields
        }
        
        return FrontMatter(**filtered_metadata)
    
    def _convert_to_html(self, content: str, front_matter: FrontMatter) -> str:
        """Convert markdown content to HTML."""
        # Convert basic markdown
        html = self.md.convert(content)
        
        # Process wiki-style links [[Page Name]] or [[slug|Display Text]]
        html = self._process_wiki_links(html)
        
        # Process image paths (mark for later media handling)
        html = self._process_image_paths(html)
        
        return html
    
    def _process_wiki_links(self, html: str) -> str:
        """Convert wiki-style links to proper HTML links."""
        # Pattern for [[Page Name]] or [[slug|Display Text]]
        wiki_pattern = r'\[\[([^\]|]+)(?:\|([^\]]+))?\]\]'
        
        def replace_wiki_link(match):
            target = match.group(1)
            display_text = match.group(2) or target
            slug = self._generate_slug(target)
            
            # Create placeholder link for later resolution
            return f'<a href="/{slug}" data-wiki-link="{target}">{display_text}</a>'
        
        return re.sub(wiki_pattern, replace_wiki_link, html)
    
    def _process_image_paths(self, html: str) -> str:
        """Process image paths and mark for later media handling."""
        # Pattern for img tags with relative paths
        img_pattern = r'<img([^>]+)src="(?!https?://)([^"]+)"([^>]*)>'
        
        def replace_image(match):
            before = match.group(1)
            src = match.group(2)
            after = match.group(3)
            
            # Mark relative images for later processing
            return f'<img{before}src="{src}" data-local-image="{src}"{after}>'
        
        return re.sub(img_pattern, replace_image, html)
    
    def _generate_slug(self, text: str) -> str:
        """Generate a URL-friendly slug from text."""
        # Convert to lowercase
        slug = text.lower()
        
        # Remove special characters and replace with hyphens
        slug = re.sub(r'[^\w\s-]', '', slug)
        slug = re.sub(r'[\s_-]+', '-', slug)
        
        # Remove leading/trailing hyphens
        slug = slug.strip('-')
        
        return slug
    
    def _detect_type_from_path(self, file_path: str) -> str:
        """Detect content type from file path."""
        path_lower = file_path.lower().replace('\\', '/')
        
        if '/podcast' in path_lower or '/episodes' in path_lower:
            return 'podcast'
        elif '/event' in path_lower or '/gathering' in path_lower:
            return 'event'
        elif '/blog' in path_lower or '/posts' in path_lower:
            return 'blog'
        elif '/page' in path_lower or path_lower.endswith('/about.md') or path_lower.endswith('/contact.md'):
            return 'page'
        
        # Default to blog
        return 'blog'
    
    def _normalize_date(self, date_value: Any) -> str:
        """Normalize date to ISO format."""
        if not date_value:
            return datetime.now().isoformat()
        
        if isinstance(date_value, str):
            try:
                # Try parsing different formats
                for fmt in ['%Y-%m-%d', '%Y-%m-%dT%H:%M:%SZ', '%Y-%m-%dT%H:%M:%S']:
                    try:
                        dt = datetime.strptime(date_value, fmt)
                        return dt.isoformat()
                    except ValueError:
                        continue
                
                # If no format works, return as-is
                return date_value
            except Exception:
                return datetime.now().isoformat()
        
        return str(date_value)

File: src/core/parser.ts
import fs from 'fs';
import path from 'path';
import matter from 'gray-matter';
import MarkdownIt from 'markdown-it';
import { glob } from 'glob';
import { MarkdownFile, FrontMatter } from '../types';
import logger from '../utils/logger';

export class MarkdownParser {
  private md: MarkdownIt;
  
  constructor() {
    this.md = new MarkdownIt({
      html: true,
      linkify: true,
      typographer: true,
      breaks: true
    });
  }
  
  async findMarkdownFiles(inputPath: string, pattern: string = '**/*.md'): Promise<string[]> {
    try {
      const fullPath = path.resolve(inputPath);
      const globPattern = path.join(fullPath, pattern);
      
      const files = await glob(globPattern, {
        ignore: ['**/node_modules/**', '**/dist/**', '**/build/**'],
        absolute: true
      });
      
      logger.info(`Found ${files.length} markdown files in ${inputPath}`);
      return files;
    } catch (error) {
      logger.error('Error finding markdown files', { error: (error as Error).message });
      throw error;
    }
  }
  
  parseFile(filePath: string): MarkdownFile {
    try {
      const content = fs.readFileSync(filePath, 'utf8');
      const parsed = matter(content);
      
      // Extract and validate front matter
      const frontMatter = this.normalizeFrontMatter(parsed.data as FrontMatter, filePath);
      
      // Convert markdown to HTML
      const html = this.convertMarkdownToHtml(parsed.content, frontMatter);
      
      return {
        path: filePath,
        frontMatter,
        content: parsed.content,
        html
      };
    } catch (error) {
      logger.error(`Error parsing file ${filePath}`, { error: (error as Error).message });
      throw error;
    }
  }
  
  private normalizeFrontMatter(data: any, filePath: string): FrontMatter {
    // Generate slug from filename if not provided
    if (!data.slug) {
      const basename = path.basename(filePath, '.md');
      data.slug = this.generateSlug(basename);
    }
    
    // Detect type from path if not specified
    if (!data.type) {
      data.type = this.detectTypeFromPath(filePath);
    }
    
    // Ensure arrays for array fields
    const arrayFields = ['authors', 'tags', 'categories', 'initiatives', 'guests'];
    arrayFields.forEach(field => {
      if (data[field] && !Array.isArray(data[field])) {
        data[field] = [data[field]];
      }
    });
    
    // Normalize dates to ISO format
    if (data.date_published) {
      data.date_published = this.normalizeDate(data.date_published);
    }
    if (data.date_updated) {
      data.date_updated = this.normalizeDate(data.date_updated);
    }
    if (data.start_date) {
      data.start_date = this.normalizeDate(data.start_date);
    }
    if (data.end_date) {
      data.end_date = this.normalizeDate(data.end_date);
    }
    
    // Set external_id if not provided
    if (!data.external_id) {
      data.external_id = data.slug;
    }
    
    return data as FrontMatter;
  }
  
  private convertMarkdownToHtml(markdown: string, frontMatter: FrontMatter): string {
    let html = this.md.render(markdown);
    
    // Handle wiki-style links [[Page Name]] or [[slug|Display Text]]
    html = this.processWikiLinks(html);
    
    // Process image paths
    html = this.processImagePaths(html, frontMatter);
    
    return html;
  }
  
  private processWikiLinks(html: string): string {
    // Pattern for [[Page Name]] or [[slug|Display Text]]
    const wikiLinkPattern = /\[\[([^\]|]+)(?:\|([^\]]+))?\]\]/g;
    
    return html.replace(wikiLinkPattern, (match, target, displayText) => {
      const slug = this.generateSlug(target);
      const text = displayText || target;
      
      // For now, create a placeholder link that will be resolved later
      return `<a href="/${slug}" data-wiki-link="${target}">${text}</a>`;
    });
  }
  
  private processImagePaths(html: string, frontMatter: FrontMatter): string {
    // Convert relative image paths to absolute paths
    // This will be updated later when images are uploaded to WordPress
    const imgPattern = /<img[^>]+src="([^"]+)"/g;
    
    return html.replace(imgPattern, (match, src) => {
      if (src.startsWith('http://') || src.startsWith('https://')) {
        return match; // Already absolute
      }
      
      // Mark relative images for later processing
      return match.replace(src, `data-local-image="${src}" src="${src}"`);
    });
  }
  
  private generateSlug(text: string): string {
    return text
      .toLowerCase()
      .trim()
      .replace(/[^\w\s-]/g, '') // Remove special characters
      .replace(/\s+/g, '-')      // Replace spaces with hyphens
      .replace(/-+/g, '-')       // Replace multiple hyphens with single
      .replace(/^-+|-+$/g, '');  // Remove leading/trailing hyphens
  }
  
  private detectTypeFromPath(filePath: string): 'blog' | 'podcast' | 'event' | 'page' {
    const normalizedPath = filePath.toLowerCase().replace(/\\/g, '/');
    
    if (normalizedPath.includes('/podcast') || normalizedPath.includes('/episodes')) {
      return 'podcast';
    }
    if (normalizedPath.includes('/event') || normalizedPath.includes('/residenc')) {
      return 'event';
    }
    if (normalizedPath.includes('/page') || normalizedPath.includes('/learn') || 
        normalizedPath.includes('/initiative')) {
      return 'page';
    }
    
    // Default to blog
    return 'blog';
  }
  
  private normalizeDate(date: any): string {
    if (!date) return new Date().toISOString();
    
    try {
      const d = new Date(date);
      if (isNaN(d.getTime())) {
        logger.warn(`Invalid date: ${date}, using current date`);
        return new Date().toISOString();
      }
      return d.toISOString();
    } catch (error) {
      logger.warn(`Error parsing date: ${date}`, { error: (error as Error).message });
      return new Date().toISOString();
    }
  }
  
  async parseFiles(files: string[]): Promise<MarkdownFile[]> {
    const results: MarkdownFile[] = [];
    
    for (const file of files) {
      try {
        const parsed = this.parseFile(file);
        results.push(parsed);
      } catch (error) {
        logger.error(`Failed to parse ${file}`, { error: (error as Error).message });
      }
    }
    
    return results;
  }
}

export default new MarkdownParser();